lib/core/src/Cardano/Wallet/Primitive/Types/TokenMap.hs

```diff
@@ -0,0 +1,529 @@
+-- its associated quantities will be added together in the resultant map.
+--
+fromFlatList :: [(AssetId, TokenQuantity)] -> TokenMap
+fromFlatList = F.foldl' acc empty
+  where
+    acc b (asset, quantity) = adjustQuantity b asset (<> quantity)
+
+-- | Creates a token map from a nested list.
+--
+-- If a token name appears more than once in the list under the same policy,
+-- its associated quantities will be added together in the resultant map.
+--
+fromNestedList
+    :: [(TokenPolicyId, NonEmpty (TokenName, TokenQuantity))] -> TokenMap
+fromNestedList entries = fromFlatList
+    [ (AssetId policy token, quantity)
+    | (policy, tokenQuantities) <- entries
+    , (token, quantity) <- NE.toList tokenQuantities
+    ]
+
+--------------------------------------------------------------------------------
+-- Deconstruction
+--------------------------------------------------------------------------------
+
+-- | Converts a token map to a flat list.
+--
+toFlatList :: TokenMap -> [(AssetId, TokenQuantity)]
+toFlatList b =
+    [ (AssetId policy token, quantity)
+    | (policy, tokenQuantities) <- toNestedList b
+    , (token, quantity) <- NE.toList tokenQuantities
+    ]
+
+-- | Converts a token map to a nested list.
+--
+toNestedList
+    :: TokenMap -> [(TokenPolicyId, NonEmpty (TokenName, TokenQuantity))]
+toNestedList =
+    fmap (fmap NEMap.toList) . Map.toList . unTokenMap
+
+--------------------------------------------------------------------------------
+-- Arithmetic
+--------------------------------------------------------------------------------
+
+-- | Adds one token map to another.
+--
+add :: TokenMap -> TokenMap -> TokenMap
+add a b = F.foldl' acc a $ toFlatList b
+  where
+    acc c (asset, quantity) =
+        adjustQuantity c asset (`TQ.add` quantity)
+
+--------------------------------------------------------------------------------
+-- Tests
+--------------------------------------------------------------------------------
+
+-- | Returns true if and only if the given map is empty.
+--
+isEmpty :: TokenMap -> Bool
+isEmpty = (== empty)
+
+-- | Returns true if and only if the given map is not empty.
+--
+isNotEmpty :: TokenMap -> Bool
+isNotEmpty = (/= empty)
+
+--------------------------------------------------------------------------------
+-- Quantities
+--------------------------------------------------------------------------------
+
+-- | Gets the quantity associated with a given asset.
+--
+-- If the given map does not have an entry for the specified asset, this
+-- function returns a value of zero.
+--
+getQuantity :: TokenMap -> AssetId -> TokenQuantity
+getQuantity (TokenMap m) (AssetId policy token) =
+    fromMaybe TQ.zero $ NEMap.lookup token =<< Map.lookup policy m
+
+-- | Updates the quantity associated with a given asset.
+--
+-- If the given quantity is zero, the resultant map will not have an entry for
+-- the given asset.
+--
+setQuantity :: TokenMap -> AssetId -> TokenQuantity -> TokenMap
```

Comment:
I just grepped for this symbol, and couldn't find where it was used by the code. Is it used by the code?

Reply:
> I just grepped for this symbol, and couldn't find where it was used by the code. Is it used by the code?

You're right that this function isn't currently used.

However, I'm almost certain that MA coin selection will need this functionality (which is why I put it in the [design document](https://jira.iohk.io/browse/ADP-347)).

For example:

When randomly selecting entries from the UTxO set, if we search for entries containing some token **_t1_**, it might be that the only entries available contain token types that don't appear in our output set.

When generating change bundles, we will need to redistribute these "also-selected" token quantities to our generated change bundles, in which case we will need some way of setting the quantity of a particular token within a token bundle.

Having said that, we could get away with just using `adjustQuantity` in this case, as `adjustQuantity (+ n)` is the same as `setQuantity n` (provided the original value is `0`).

@rvl if you prefer, I can remove it and re-add it in a later PR that addresses coin selection?

Reply:
Yes please - I am looking forward to reviewing this function in the _later_ PR.

Reply:
@rvl My apologies: I had forgotten that `setQuantity` is used by `adjustQuantity`.

That function is used by `add` and `fromFlatList`, which is used by the DB deserialization functions.

So it's not really possible to remove without removing functionality that we need.

Reply:
OK.